### ISSUE TITLE
F-Droid build for 0.5.5 failed

### ISSUE THREAD

**Issue by mvglasow** [navit-gps/navit#1085]
**TL; DR: I see the 0.5.5 build never made it on F-Droid. Based on my analysis, it seems the build fails on an assertion. The CircleCI build of the same version was successful, though. I am wondering if this could be caused by an incompatible NDK version.**

The F-Droid metadata for 0.5.5 has:

    disable: fails to build

The build was attempted; a truncated log is at https://f-droid.org/wiki/page/org.navitproject.navit/lastbuild_2008082307. It shows some error occurred while building the C code, though the actual error is not visible in the excerpt. (On a side note, we should really reduce the number of compiler warnings—either fix the code or disable warnings we don’t consider an issue.) As jochensp on F-Droid IRC tells me, the original log is probably no longer available.

However, our own CI was able to build Navit (it still shows as failed, as the artifact upload failed later on), see https://github.com/navit-gps/navit/runs/962451579.

I forked the F-Droid metadata repo and tried to trigger a build of Navit on their CI chain. In order for this to work, I had to remove the line

    ndk: r20b

from the metadata, as this would break CI—the CI image only has NDK r12b (the live build server has a wider choice of versions), and anyways gradle will download the NDK if no suitable installation is available (in this case NDK 22 got installed).

The build fails with the following error while running `clang` on `wordexp.c` for the `armv7-none-linux-androideabi16` target:

```
  In file included from /builds/mvglasow/fdroiddata/build/org.navitproject.navit/navit/support/wordexp/wordexp.c:15:
  In file included from /usr/include/glib-2.0/glib.h:30:
  In file included from /usr/include/glib-2.0/glib/galloca.h:32:
  /usr/include/glib-2.0/glib/gtypes.h:423:3: error: '_GStaticAssertCompileTimeAssertion_0' declared as an array with a negative size
    G_STATIC_ASSERT(sizeof (unsigned long long) == sizeof (guint64));
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /usr/include/glib-2.0/glib/gmacros.h:232:103: note: expanded from macro 'G_STATIC_ASSERT'
  #define G_STATIC_ASSERT(expr) typedef char G_PASTE (_GStaticAssertCompileTimeAssertion_, __COUNTER__)[(expr) ? 1 : -1] G_GNUC_UNUSED
                                                                                                        ^~~~~~~~~~~~~~~
```

(For the full log, see: https://gitlab.com/mvglasow/fdroiddata/-/jobs/1012058113#L10583)

Apparently, the assertion that `sizeof (unsigned long long) == sizeof (guint64)` fails (those two types should be the same size but they are not).

A [StackOverflow post on a similar issue in a different project](https://stackoverflow.com/a/39908759/2703209) indicates that this might be caused by an architecture mismatch, i.e. building for architecture A and trying to include a library built for architecture B. I am wondering if it could also be caused by making incorrect assumptions about the target architecture somewhere in the code or toolchain. What strikes me is that we’re apparently expecting to use 64-bit integers on a 32-bit architecture.

I retried the F-Droid CI build with current master (37f22db) and got the same error. However, CI passed (see https://github.com/navit-gps/navit/runs/1801436556) and the build succeeds on my own machine as well.

GitLab CI downloads and uses NDK 22; our CI has NDK 20, my local machine runs NDK 18.

I had downgraded from NDK 21 on my local machine because of an error—after reactivating it, the local build of 37f22db fals with:
```
  /home/michael/src/navit/navit/support/gettext_intl/dcigettext.c:160:7: error: redeclaration of 'getcwd' must have the 'overloadable' attribute
  char *getcwd ();
        ^
  /home/michael/tools/android-sdk-linux_86/ndk/21.0.6113669/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/bits/fortify/unistd.h:105:7: note: previous overload of function is here
  char* getcwd(char* const __pass_object_size buf, size_t size)
        ^
```

So I am wondering if an incompatible NDK is the culprit here. OTOH, the live F-Droid build server should have used NDK 20, which is the same we use successfully on our CI environment.

**Comment by mvglasow** [comment]
Just tried NDK 22 on my local machine, which also fails with the `getcwd` overload error. I’m stuck here, could somebody with more profound toolchain knowledge than myself take a look at this? @jkoan maybe?

**Comment by mvglasow** [comment]
Regarding the error on NDK 21 and 22, see #1043

**Comment by mvglasow** [comment]
#792 has a full log of the failed build on the live server; it shows the same error I am getting on CI (failed assertion).

**Comment by mvglasow** [comment]
From IRC:

```
<jkoan> i have a suspission that this doesn't match up: /builds/mvglasow/fdroiddata/... and /usr/include/glib-2.0 the android build should use the glib from the ndk not from the host
<mvglasow> then I wonder why that works on our end but not on F-Droid
[...]
<jkoan> yes, me to
<jkoan> will try to figure this out
<jkoan> probably because fdroid will run gradle again after our build_android.sh is done? (just a guess for now)
```

Digging a bit deeper, I also found some warnings regarding CMake compiler identification not being available on versions below 3.10 (see https://gitlab.com/mvglasow/fdroiddata/-/jobs/1012058113#L4937). However, while CMake 3.10 helped get rid of these warnigs, it still produced the same error.

The first lines of CMake output for the target platforms on CI indicate that glib 2.50.3 was found (which is probably the host version). On my local system none was found, hence CMake reverted to the support version in Navit (as is evident from subsequent build warnings, which refer to files in support/glib). The same seems to happen on CircleCI—no glib is found, therefore it uses the bundled support library.

Synaptic tells me I have libglib2.0-dev, version 2.64.6.1~ubuntu20.04.1 installed locally, so it can’t be an issue of libglic2.0-dev being installed on one machine but not on the other.

So I can confirm the suspicion @jkoan mentioned: CMake is trying to use a host library while cross-compiling. Because the host architecture is 64 bit and the target architecture is 32 bit, the check above fails. However, the correct library to use would be the support library, not something bundled with the NDK. Also, as for F-Droid re-running gradle after our script finishes, it looks to me as if the error occurs on the first run of gradle.

The question remains: why does CMake pick up a host library for cross-compilation on GitLab CI but not when running locally or on our own CI? Apparently something’s confusing CMake’s logic for detecting the availability of certain libraries, though I am not very familiar with that kind of toolchain details.

Detection of glib seems to happen in `./CMakeLists.txt` lines 158–211: First `find_package(PkgConfig)` is called to find out if pkg-config is available. If so, glib presence is verified with `pkg_check_modules (GLIB2 glib-2.0>=2.10)`. If either pkg-config or glib are missing, `support/glib` is enabled (default is disabled).

Output from `pkg_check_modules` (on the builds that work) tells me that pkg-config was found and `pkg_check_modules` reported the respective library was not found.

I came across a related question on StackOverflow about incorrect pkg-config detections when cross-compiling: https://stackoverflow.com/questions/9221236/pkg-config-fails-to-find-package-under-sysroot-directory. I see our build script clears the `PKG_CONFIG_LIBDIR` environment variable if it is set. @jkoan, maybe you can make some sense out of the SO post. I might still run some tests to see what is in the variables, but not today…

**Comment by mvglasow** [comment]
Adding the following lines to CMakeLists.txt:
```
message(STATUS "  CMAKE_SYSROOT: $ENV{CMAKE_SYSROOT}")
message(STATUS "  PKG_CONFIG_DIR: $ENV{PKG_CONFIG_DIR}")
message(STATUS "  PKG_CONFIG_LIBDIR: $ENV{PKG_CONFIG_LIBDIR}")
message(STATUS "  PKG_CONFIG_SYSROOT_DIR: $ENV{PKG_CONFIG_SYSROOT_DIR}")
```
tells me all of these vars are empty. Basically, it seems pkg-config is picking up the wrong libraries despite the variables being empty, but I’ve been unable to figure out why.

Just as a test, I changed the condition for the whole pkg-config detection block (around line 170) to
```
if(PKG_CONFIG_FOUND AND NOT ANDROID)
```
i.e. skipping the detection when building for Android. (I also modified the compiler warning flags when building for Android to work around #1086.) That gets me one step further: the source compiles but eventually the linker reports:
```
  ld: error: undefined symbol: __memcpy_chk
  >>> referenced by string.h:48 (bionic/libc/include/bits/fortify/string.h:48)
  >>>               crc_folding.o:(crc_fold_copy) in archive /opt/android-sdk/ndk/22.0.7026061/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/16/libz.a
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
  ninja: build stopped: subcommand failed.
```

**Comment by mvglasow** [comment]
Disabling `add_compiler_flag_if_available("-D_FORTIFY_SOURCE=2")` for Android had zero effect on the last error. Anyway, I feel all this `if (NOT ANDROID)` magic in CMake is just a workaround around something that’s broken at the core. We need to find out why CMake is trying to build against host libraries (just on this CI, not on another) and fix that.

**Comment by mvglasow** [comment]
Just noticed that the failed linker command is passed an argument, `--sysroot=/opt/android-sdk/ndk/22.0.7026061/toolchains/llvm/prebuilt/linux-x86_64/sysroot`. As I understand, sysroot is the path in which to look for libraries. Also, the error is generated by a header file which seems to be part of bionic, Android’s equivalent to glibc (mostly for license reasons afaik, glibc is under a GNU license while bionic is Apache licensed).

So apparently the build is not entirely consistent in its choice of C suport library, which is causing issues here. Though I still don’t understand why this works on our CI Image but not on F-Droid’s.

It seems every NDK version has one or two `sysroot` dirs:
* `$NDK/sysroot` and
* `$NDK/toolchains/llvm/prebuilt/linux-x86_64/sysroot` (starting somewhere around NDK 19–21)

We could try passing one of these two to CMake (with pkg-config re-enabled) and see if that changes anything.

**Comment by mvglasow** [comment]
Since I noticed earlier that I am using NDK 18.1.5063045 for my local build, whereas our CI image has 20 and the F-Droid image auto-installs 22.0.7026061, I tried forcing the build to use that particular NDK version. That, together with the `if(PKG_CONFIG_FOUND AND NOT ANDROID)` hack, finally led to a build which successfully generates an APK file. However, even on NDK 20, if I re-enable pkg-config, it starts building against the wrong library again.

**Comment by mvglasow** [comment]
for the artifact see https://gitlab.com/mvglasow/fdroiddata/-/jobs/1044508536/artifacts/browse/unsigned/

**Comment by mvglasow** [comment]
The remaining obstacle to passing NDK’s sysroot to CMake is finding its installation dir. @jkoan wrote on IRC:

>  I think somewhere within gradle must be a variable where gradle knows the ndk path. We could then pass this to cmake as a parameter and set the environment variable from there before pkg-config is called

**Comment by jkoan** [comment]
> The remaining obstacle to passing NDK’s sysroot to CMake is finding its installation dir. @jkoan wrote on IRC:
> 
> > I think somewhere within gradle must be a variable where gradle knows the ndk path. We could then pass this to cmake as a parameter and set the environment variable from there before pkg-config is called

According to [this](https://stackoverflow.com/questions/32751398/how-can-i-access-sdk-ndk-paths-in-my-gradle-build-scripts) StackOverflow Question the ndk path is within the variable `android.ndkDirectory` which i can confirm.
So the next step is to plug everything into the build.gradle and CMakeLists and hope for the best

**Comment by jkoan** [comment]
> Adding the following lines to CMakeLists.txt:
> 
> ```
> message(STATUS "  CMAKE_SYSROOT: $ENV{CMAKE_SYSROOT}")
> message(STATUS "  PKG_CONFIG_DIR: $ENV{PKG_CONFIG_DIR}")
> message(STATUS "  PKG_CONFIG_LIBDIR: $ENV{PKG_CONFIG_LIBDIR}")
> message(STATUS "  PKG_CONFIG_SYSROOT_DIR: $ENV{PKG_CONFIG_SYSROOT_DIR}")
> ```
> 
> tells me all of these vars are empty. Basically, it seems pkg-config is picking up the wrong libraries despite the variables being empty, but I’ve been unable to figure out why.

Probably another difference is between null and empty, Probably the Environment Variables are unset instead of set to a string with zero length

**Comment by mvglasow** [comment]
> According to [this](https://stackoverflow.com/questions/32751398/how-can-i-access-sdk-ndk-paths-in-my-gradle-build-scripts) StackOverflow Question the ndk path is within the variable `android.ndkDirectory` which i can confirm.

Any idea how I can pass this to cmake?

**Comment by mvglasow** [comment]
#1097 fixes the F-Droid build, and #1098 has added a CI job for F-Droid, thus a working F-Droid build is now required to pass CI. If code changes require a change to the F-Droid build recipe, this would now be noticed immediately and could be tested; changes to the local build recipe can then be applied to the one in the F-Droid repo.

Changes to the F-Droid build recipe were necessary and have been submitted to F-Droid as https://gitlab.com/fdroid/fdroiddata/-/merge_requests/8472; I am leaving this ticket open until the MR is through.

**Comment by mvglasow** [comment]
The MR us through and we have a new version on F-Droid. Closing.
